diff --git a/package/openwrt/files/etc/init.d/smartdns b/package/openwrt/files/etc/init.d/smartdns
index 83c5366..1876c63 100644
--- a/package/openwrt/files/etc/init.d/smartdns
+++ b/package/openwrt/files/etc/init.d/smartdns
@@ -286,6 +286,7 @@ load_server()
 
 restart_crond()
 {
+	[ -n "$SMARTDNS_UPDATING_FILES" ] && return
 	/etc/init.d/cron restart >/dev/null 2>&1
 }
 
@@ -308,7 +309,7 @@ disable_auto_update()
 
 enable_auto_update()
 {
-	grep "0 $auto_update_day_time * * $auto_update_week_time /etc/init.d/smartdns updatefiles" /etc/crontabs/root 2>/dev/null
+	grep -F "0 $auto_update_day_time * * $auto_update_week_time /etc/init.d/smartdns updatefiles" /etc/crontabs/root 2>/dev/null
 	if [ $? -eq 0 ]; then
 		return 
 	fi
@@ -1074,6 +1075,7 @@ check_and_add_entry() {
 }
 
 updatefiles() {
+	export SMARTDNS_UPDATING_FILES=1
 	config_load "smartdns"
 	[ ! -d "$SMARTDNS_DOWNLOAD_TMP_DIR" ] && mkdir -p "$SMARTDNS_DOWNLOAD_TMP_DIR"
 	config_foreach download_file "download-file"
