--- a/netfn/Makefile
+++ b/netfn/Makefile
@@ -10,8 +10,8 @@ OBJECTS = $(SOURCES:$(SRCDIR)/%.c=$(OBJ
 
 INCLUDE += -I./include
 EXTRA_CFLAGS = -Wall -Werror -fpie -Wl,-z,relro -Wl,-z,now
-EXTRA_LDFLAGS = -lnl-3 -lnl-genl-3 -pie -Wl,-z,relro -Wl,-z,now
-LDLIBS += -ljansson
+EXTRA_LDFLAGS = -lnl-tiny -pie -Wl,-z,relro -Wl,-z,now
+LDLIBS += -ljansson
 
 all: release
 
--- a/netfn/netfn_rule.c
+++ b/netfn/netfn_rule.c
@@ -96,7 +96,7 @@ int netfn_rule_config(struct netfn_rule_
 	/*
 	 * send netlink msg
 	 */
-	err = nl_send_sync(nrs->sk, msg);
+	err = nl_send_auto_complete(nrs->sk, msg);
 	if (err < 0) {
 		netfn_log_error("%pFailed to send nl message, family(%s) with error(%d)\n", nrs, family, err);
 		goto fail;
--- a/ppe/ppenl_lib/nss_ppenl_sock.c
+++ b/ppe/ppenl_lib/nss_ppenl_sock.c
@@ -282,7 +282,7 @@ int nss_ppenl_sock_send(struct nss_ppenl
 	 * In case firmware response is sent before nl_recvmsgs is invoked,
 	 * the response will be queued until the listener is available.
 	 */
-	error = nl_send_sync(sock->nl_sk, msg);
+	error = nl_send_auto_complete(sock->nl_sk, msg);
 	if (error < 0) {
 		nss_ppenl_sock_log_error("%d:failed to send (family:%s, error:%d)\n", pid, sock->family_name, error);
 		nss_ppenl_sock_deref(sock);
--- a/ppe/ppecfg/Makefile
+++ b/ppe/ppecfg/Makefile
@@ -8,11 +8,11 @@ SOURCES = $(wildcard $(SRCDIR)/*.c)
 HEADERS = $(wildcard $(SRCDIR)/*.h)
 OBJECTS = $(SOURCES:$(SRCDIR)/%.c=$(OBJPATH)/%.o)
 
-INCLUDE += -I./include
+INCLUDE += -I./include -I../ppenl_lib/include
 EXTRA_CFLAGS = -Wall -Werror -fpie -Wl,-z,relro -Wl,-z,now
-EXTRA_LDFLAGS = -lnl-genl-3 -lnl-ppe -pie -Wl,-z,relro -Wl,-z,now
-LDLIBS  = -L../ppecfg_lib/obj
-LDLIBS  += -ljson-c
+EXTRA_LDFLAGS = -lnl-ppe -lnl-tiny -pie -Wl,-z,relro -Wl,-z,now
+LDLIBS  = -L../ppenl_lib/obj
+LDLIBS  += -ljson-c
 
 all: release
 
