diff -Naru ghostscript-10.0.0_orig/configure ghostscript-10.0.0/configure
--- ghostscript-10.0.0_orig/configure	2022-09-21 17:39:46.000000000 +0900
+++ ghostscript-10.0.0/configure	2022-09-23 16:32:58.866147602 +0900
@@ -9477,6 +9477,9 @@
 JB2_STDINT_TYPES_IN=
 JBIG2DEC_REQ=0.19
 
+# force using local library
+with_jbig2dec=yes
+
 if test "x$JBIG2_DECODER" = x; then
 
 # Check whether --with-jbig2dec was given.
@@ -11072,6 +11075,8 @@
 PDF_MAK="\$(GLSRCDIR)\$(D)stub.mak"
 PDFROMFS_MAK="\$(GLSRCDIR)\$(D)stub.mak"
 
+JPX_DECODER=openjpeg
+
 if test x"$with_pdf" != x"no" ; then
 
   if test x"$JBIG2_DECODER" = x""; then
