Index: cups-2.4.2/config-scripts/cups-directories.m4
===================================================================
--- cups-2.4.2.orig/config-scripts/cups-directories.m4
+++ cups-2.4.2/config-scripts/cups-directories.m4
@@ -96,7 +96,7 @@ dnl Fix "libdir" variable...
 AS_IF([test "$libdir" = "\${exec_prefix}/lib"], [
     AS_CASE(["$host_os_name"], [linux*], [
 	AS_IF([test -d /usr/lib64 -a ! -d /usr/lib64/fakeroot], [
-	    libdir="$exec_prefix/lib64"
+	    libdir="$exec_prefix/lib"
 	], [
 	    libdir="$exec_prefix/lib"
 	])
Index: cups-2.4.2/configure
===================================================================
--- cups-2.4.2.orig/configure
+++ cups-2.4.2/configure
@@ -7302,7 +7302,7 @@ then :
 	if test -d /usr/lib64 -a ! -d /usr/lib64/fakeroot
 then :
 
-	    libdir="$exec_prefix/lib64"
+	    libdir="$exec_prefix/lib"
 
 else $as_nop
 
Index: cups-2.4.2/cups-config.in
===================================================================
--- cups-2.4.2.orig/cups-config.in
+++ cups-2.4.2/cups-config.in
@@ -44,7 +44,7 @@ else
 	CFLAGS="$CFLAGS -I$includedir"
     fi
 
-    if test $libdir != /usr/lib -a $libdir != /usr/lib32 -a $libdir != /usr/lib64; then
+    if test $libdir != /usr/lib -a $libdir != /usr/lib32 -a $libdir != /usr/lib; then
 	LDFLAGS="$LDFLAGS -L$libdir"
     fi
 fi
