--- a/crates/shadowsocks/Cargo.toml
+++ b/crates/shadowsocks/Cargo.toml
@@ -28,13 +28,11 @@ trust-dns = ["hickory-dns"]
 # Users should always avoid using these ciphers in practice
 stream-cipher = [
     "shadowsocks-crypto/v1-stream",
-    "shadowsocks-crypto/aws-lc",
 ]
 
 # Enable AEAD ciphers
 aead-cipher = [
     "shadowsocks-crypto/v1-aead",
-    "shadowsocks-crypto/aws-lc",
 ]
 
 # Enable extra AEAD ciphers
@@ -43,7 +41,6 @@ aead-cipher-extra = ["aead-cipher", "shadowsocks-crypto/v1-aead-extra"]
 
 aead-cipher-2022 = [
     "shadowsocks-crypto/v2",
-    "shadowsocks-crypto/aws-lc",
     "rand",
     "aes",
     "lru_time_cache",

